driver/src/main/com/mongodb/GridFSDownloadStreamImpl.java

```diff
@@ -0,0 +1,171 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb;
+
+import com.mongodb.client.GridFSDownloadStream;
+import com.mongodb.client.MongoCollection;
+import org.bson.BsonDocument;
+import org.bson.BsonInt32;
+import org.bson.BsonObjectId;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import static com.mongodb.assertions.Assertions.isTrue;
+import static com.mongodb.assertions.Assertions.notNull;
+import static java.lang.String.format;
+
+class GridFSDownloadStreamImpl extends InputStream implements GridFSDownloadStream {
+    private final BsonDocument fileInfo;
+    private final MongoCollection<BsonDocument> chunksCollection;
+    private final BsonObjectId fileId;
+    private final long lengthInBytes;
+    private final int chunkSizeInBytes;
+    private final int numberOfChunks;
+    private int chunkIndex;
+    private int bufferOffset;
+    private long currentPosition;
+    private byte[] buffer = null;
+    private boolean closed;
+
+    GridFSDownloadStreamImpl(final BsonDocument fileInfo, final MongoCollection<BsonDocument> chunksCollection) {
+        this.fileInfo = notNull("file information", fileInfo);
+        this.chunksCollection = notNull("chunks collection", chunksCollection);
+
+        try {
+            isTrue("file information contains _id", fileInfo.containsKey("_id") && fileInfo.get("_id").isObjectId());
+            isTrue("file information contains length", fileInfo.containsKey("length") && fileInfo.get("length").isInt64());
+            isTrue("file information contains chunkSize", fileInfo.containsKey("chunkSize") && fileInfo.get("chunkSize").isInt32());
+        } catch (IllegalStateException e) {
+            throw new MongoGridFSException("GridFS file information is not in the expected format", e);
+        }
+        fileId = fileInfo.getObjectId("_id");
+        lengthInBytes = fileInfo.getInt64("length").getValue();
+        chunkSizeInBytes = fileInfo.getInt32("chunkSize").getValue();
+        numberOfChunks = (int) Math.ceil((double) lengthInBytes / chunkSizeInBytes);
+    }
+
+    @Override
+    public BsonDocument getFileInformation() {
+        return fileInfo;
+    }
+
+    @Override
+    public InputStream getInputStream() {
+        return this;
+    }
+
+    @Override
+    public int read() throws IOException {
+        byte[] b = new byte[1];
+        int res = read(b);
+        if (res < 0) {
+            return -1;
+        }
+        return b[0] & 0xFF;
+    }
+
+    @Override
+    public int read(final byte[] b, final int off, final int len) throws IOException {
+        checkClosed();
+
+        if (currentPosition == lengthInBytes) {
```

Comment:
Reads funny to me because one variable has an InBytes suffix and the other doesn't.  I think the suffix can safely be removed from the fields that have them without affecting clarity.